ships/0024-surfacing-errors-to-buildrun.md

```diff
@@ -117,11 +117,56 @@ As a third-party application developer that uses shipwright under the hood, I wa
 
 ### Implementation Notes
 
-`tbd in the next phase of the proposal`
+
+This ship proposes two changes in the API that are implemented in `buildrun_types.go` and thus propagated in the CRD. First, the BuildRun's status gets the new field `errorDetails` that embeds the `failedAt` field:
+
+```go
+type FailureDetails struct {
+    Reason   string    `json:"reason,omitempty"`
+    Message  string    `json:"message,omitempty"`
+    Location *FailedAt `json:"location,omitempty"`
+}
+``` 
+
+Second, we keep the `failedAt` field to prevent breaking user applications. However, it should be deprecated and removed in the future due to its redundancy. Therefore, we propose adding documents that state the deprecation status and adding the news to the release notes.
+To ease deprecation and future removal, we propose refactoring `UpdateBuildRunUsingTaskRunCondition` by extracting the logic of finding the failed pod and container in a BuildRun/TaskRun.
+The new file `pkg/reconciler/buildrun/resources/failures.go` includes the extraced logic and other failure related functions.
+
+A TaskRun can emit failures using two results:
+1. shp-error-reason
+2. shp-error-message
+
+Whenever a Tekton TaskRun fails, it keeps a JSON string of all results inside each terminated step in `TaskRun.Status.Steps[i].Terminated.Message` of type `PipelineResourceResult`. Whenever we can find the two error results in a step's message field, we surface the error as `failureDetails`.
+
+So far, we have presented a mechanism to lift errors results into a BuildRun. However, no component has produced error results yet.  One source of failure is misconfiguration in the source step for git applications. For example, a user can forget to provide authentication details for a private repository. Therefore, we propose to implement error results for `cmd/git`. On top of that, we propose implementing smarter error reporting by parsing and classifying git's messages for errors in `stdout`. A third-party application would be able to provide mitigations for failed builds based on error classes in git, for example, a UI workflow to reenter basic authentication details.
+
+The following errors classes are implemented for git:
+- `AuthInvalidUserOrPass`, expresses that basic authentication is not possible
+- `AuthInvalidKey`, expresses that ssh authentication is not possible
+- `BranchNotFound` expresses that a remote branch does not exist
```

Comment:
What about `RevisionNotFound` because the field in the Build spec is called revision and can also contain a tag name or commit id ?